kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -177,6 +195,50 @@ public static int findColIndex(StructType schema, String colName) {
     return -1;
   }
 
+  ////////////////////
+  // Helper methods //
+  ////////////////////
+
+  /**
+   * Recursively checks if {@code supersetType} is equivalent to {@code subsetType} (if applicable)
+   * or if {@code supersetType} is a superset of {@code subsetType}.
+   *
+   * <ul>
+   *   <li>If both are {@code StructType}, compare fields recursively.
+   *   <li>If both are {@code ArrayType}, compare element types.
+   *   <li>If both are {@code MapType}, compare key & value types.
+   *   <li>Otherwise, use the standard {@link StructType#equivalent} check.
+   * </ul>
+   */
+  private static boolean isSupersetOrEquivalent(DataType supersetType, DataType subsetType) {
+    if (subsetType instanceof StructType) {
+      if (!(supersetType instanceof StructType)) {
+        return false;
+      }
+      StructType supersetStruct = (StructType) supersetType;
+      StructType subsetStruct = (StructType) subsetType;
+      return isSuperset(supersetStruct, subsetStruct);
```

Comment:
No need to block on this, but it's a little weird that we recurse back and forth between these two fx. Wondering if the code in `isSuperSet` (i.e. iterating through the fields) should just sit here? Seems like that makes sense and matches the pattern we follow for the other 2 complex types

Reply:
That's a good observation. The desire for these two separate methods arises from us wanting a distinct public API that just compares two `StructType`s.

For example, we wouldn't want to expose a public utility like `isSupersetOrEquivalent(DataType supersetType, DataType subsetType)`. Just exposing one that takes in a StructType seems most intuitive for me.

Reply:
I was wondering if we even need this API here in `SchemaUtils`. Do you think having `StructType::isSuperSet` is sufficient?

Reply:
I added it to try and encourage us to move away from a very c style of programming where you are just passing in structs/values into utils, and instead of use a more OOP style where you invoke methods on instances of classes.

I find the former not very discoverable and the latter to be very discoverable, in terms of what are the operations I can apply on a given struct/class?

open to thoughts!

Reply:
I think I was asking for the opposite... I was wondering if we can just mainly have `StructType::isSuperSet` and then some more "internal" method here in the `SchemaUtils` that it calls (like this recursive `isSupersetOrEquivalent`).

I'm definitely good with having `StructType::isSuperSet`-- not asking to remove that at all

Reply:
agree the recursion between the two functions feels non-ideal. what I was thinking (and maybe this is captured above - not clear what the exact proposal is):

since there is already a public interface for `isSuperSetOf` and `isSubsetOf`, can't there be a single `isSupersetOrEquivalent` that just recurses itself? (i.e. combine this and the `isSuperset` above)